pkg/common/volume_operations.go

```diff
@@ -426,53 +426,75 @@ func (vo *VolumeOperationsImpl) UpdateCRsAfterVolumeDeletion(ctx context.Context
+		return
 	}
-	// AC CR must exist
-	if acCR.Name == "" {
-		ll.Errorf("Unable to find available capacity resource for volume %s", volumeID)
+	driveCR, err := vo.crHelper.GetDriveCRByVolume(&volumeCR)
+	if err != nil {
+		ll.Errorf("Unable to read Drive CR for Volume %s : %v", volumeCR.Name, err)
 		return
 	}
 
-	// for LogicalVolumeGroup SCs we need to delete AC CR when no volumes remain to avoid new allocations since
-	// underlying LogicalVolumeGroup CR is destroying. For other SC just to increase size
-	isDeleted := false
-	lvg := &lvgcrd.LogicalVolumeGroup{}
 	if util.IsStorageClassLVG(volumeCR.Spec.StorageClass) {
-		if err = vo.k8sClient.ReadCR(context.Background(), volumeCR.Spec.Location, "", lvg); err != nil {
-			ll.Errorf("Unable to get LogicalVolumeGroup %s: %v", volumeCR.Spec.Location, err)
+		if err = vo.updateLVGAfterVolumeDeletion(ctx, &volumeCR, driveCR, acCR); err != nil {
 			return
 		}
-
-		if isDeleted, err = vo.deleteLVGIfVolumesNotExistOrUpdate(lvg, volumeCR.Name, &acCR); err != nil {
-			ll.Errorf("Unable to remove volume reference from LogicalVolumeGroup %s: %v", volumeCR.Spec.Location, err)
-		}
 	}
 
-	// if LogicalVolumeGroup wasn't deleted and health of volume is GOOD increase AC size
-	// We don't increase AC size for unhealthy volume to avoid new allocations on top of unhealthy drive/lvg
-	if !isDeleted && volumeCR.Spec.Health == apiV1.HealthGood {
+	if volumeCR.Spec.Health == apiV1.HealthGood {
 		// Increase size of AC using volume size
 		acCR.Spec.Size += volumeCR.Spec.Size
-		if err = vo.k8sClient.UpdateCRWithAttempts(ctx, &acCR, 5); err != nil {
-			ll.Errorf("Unable to update AC %s size: %v", acCR.Name, err)
+		ll.Debugf("Add %d to size of AC %s", volumeCR.Spec.Size, acCR.Name)
+		if err := vo.k8sClient.UpdateCR(ctx, acCR); err != nil {
+			ll.Errorf("Unable to update AC %s CR: %v", acCR.Name, err)
 		}
 	}
 }
 
+func (vo *VolumeOperationsImpl) updateLVGAfterVolumeDeletion(ctx context.Context, volumeCR *volumecrd.Volume,
+	driveCR *drivecrd.Drive, acCR *accrd.AvailableCapacity) error {
+	ll := vo.log.WithFields(logrus.Fields{
+		"method":   "updateLVGAfterVolumeDeletion",
+		"volumeID": volumeCR.Name,
+	})
+
+	lvg := lvgcrd.LogicalVolumeGroup{}
+	if err := vo.k8sClient.ReadCR(ctx, volumeCR.Spec.Location, "", &lvg); err != nil {
+		ll.Errorf("Unable to get LogicalVolumeGroup %s: %v", volumeCR.Spec.Location, err)
+		return err
+	}
+
+	// If volume not last, remove from lvg ref list, If volume is last need convert AC to Drive and remove LVG
+	if len(lvg.Spec.VolumeRefs) > 1 {
+		lvg.Spec.VolumeRefs = util.RemoveString(lvg.Spec.VolumeRefs, volumeCR.Name)
+		ll.Debugf("Remove volume ref %s from LVG %s", volumeCR.Name, lvg.Name)
+		if err := vo.k8sClient.UpdateCR(ctx, &lvg); err != nil {
+			ll.Errorf("Unable to update LVG %s CR: %v", lvg.Name, err)
+			return err
+		}
+	} else {
+		ll.Debugf("Delete LVG %v", lvg.Name)
```

Comment:
In my opinion we should update AC at first and then delete LVG. As I see in drive_controller changes, we check LVG before AC deletion try, so it my couse to:
1. controller: delete LVG
2. drive_controller: check LVG 
3. drive_controller: try to delete AC and fail
4. controller: update AC

Reply:
yes, it will be changed